### ISSUE TITLE
Quantity validation should reject negative numbers for their negativity

### ISSUE THREAD

**Issue by bugfolder** [backdrop-contrib/ubercart#486]
If you enter a negative number (e.g., -1) for the quantity in an Ubercart quantity field (like at /cart), it errors with the message:

> The quantity must be an integer.

Well, it's good that it errors, but technically, -1 is an integer. So we should modify that error message. Perhaps

> The quantity must be a non-negative integer.

But since "integer" is a little bit math-ey-sounding, perhaps even better:

> The quantity must be a whole number (0, 1, 2, ...etc).

Note that this text shows up in uc_product.test, so it will need to be changed in the test to match.

**Comment by argiepiano** [comment]
Would it make sense to have a `'number'` widget instead of a textfield for quantity? This is defined in `uc_store_element_info`. With a `'number'` widget you can limit the input to integers numbers equal or higher than 0.

**Comment by bugfolder** [comment]
Yes, that's what I was thinking of. That will also make it easier for uc_clearance module to handle clearance items, because it can set the maximum value to what's in stock.

**Comment by bugfolder** [comment]
Only drawback to number widgets is their behavior is very browser-specific in terms of how they handle bad input.

**Comment by argiepiano** [comment]
> Only drawback to number widgets is their behavior is very browser-specific in terms of how they handle bad input.

True - but they are still used in Backdrop. But if you prefer to avoid them, I'd go for an error message like "The quantity must be 0 or whole number larger than 0."

**Comment by bugfolder** [comment]
> True - but they are still used in Backdrop. But if you prefer to avoid them...

Oh yes, I know; I worked on one or more core PRs related to them, which was how I encountered the browser differences. I'll try them out on this issue and see how they work. (If I don't like them, I'll explain why in the issue.)

**Comment by bugfolder** [comment]
I tried out the number type; here's some results. For point of reference, here's the behavior of the current form for an invalid value (same result for both "-1" and "1.5"):

![image](https://github.com/backdrop-contrib/ubercart/assets/29214084/e5ed430c-3b94-4d66-96a7-bace03bd5695)

Now we change it to a number field (with `'#min' => 0` and `'#step' => 1`). In Safari, here's the error with a value of -1:

![image](https://github.com/backdrop-contrib/ubercart/assets/29214084/b5f2b950-babc-43a2-925c-9343d844eb93)

Note that the `'#size' => 5` was ignored, and the field was made wider. (This can be fixed with adding CSS like `width: 5em;`.) There's no red error box, because the error is generated at the browser level, and it's pretty clear and explanatory.

Entering "1.5" generates a different browser error:

![image](https://github.com/backdrop-contrib/ubercart/assets/29214084/e6f4e5d3-21a4-4b11-87c7-26a138e8c465)

This is less helpful than an explicit message about what constitutes a "valid value."

Chrome is a little more helpful, giving suggestions, but still doesn't explain what constitutes "valid":

![image](https://github.com/backdrop-contrib/ubercart/assets/29214084/3cb19566-8ac7-413d-bd6f-d5400db421b7)

Firefox does the same thing as Chrome with different tooltip styling, but it at least respects the size (note that the box is narrower than the previous):

![image](https://github.com/backdrop-contrib/ubercart/assets/29214084/885b2836-56ea-4a15-9fc9-ebd6dbd36436)

So, I guess the question is, do we use the `number` field and live with the browser-to-browser inconsistency and ambguous error messages, or stick with `textfield` where we have greater control over the UX? I think I incline to the latter. @robertgarrigos, @laryn, @argiepiano, your thoughts?

**Comment by laryn** [comment]
I'm inclined to use `number`, reducing code maintenance and hopefully get further in-browser improvements as they come along... but would defer if others felt strongly. Does anyone know if there are benefits one way or the other in terms of accessibility?

**Comment by ProfessorGram** [comment]
This is forms validation technical debt, pure and simple.  

IMHO it's better to keep forms validation and error-handling in-team rather than rely on other teams to come up with a solution "soon", especially when the other team don't know this requirement exists, nor its priority.

Why not just develop better form validation and input-handling in the form of a globally available function that:

1)  Accepts text input in any form
2)  Sanitizes it to prevent hacking (removing special characters, etc...)
3)  Validates it.  If the input is not somehow acceptable, explaining why in an error message
4)  Returns a "safe" value that is either a positive integer (i.e. a  "Natural Number") or ZERO
5)  If the return value is ZERO, the informational message explains things.
6)  If the return value is a positive integer, use it.

Am I missing something?

https://stackoverflow.com/questions/4878194/regex-to-validate-only-natural-numbers

https://www.w3schools.com/php/php_numbers.asp#:~:text=PHP%20Numerical%20Strings,whether%20a%20variable%20is%20numeric.

**Comment by bugfolder** [comment]
After all this discussion (and over a year), simply changing the message to `"The quantity must be zero or a positive whole number."` appears to address the issue, so I've created a PR with that change.
